### PR TITLE
[Tabs] Adds safe area inset

### LINKED ISSUE #7733: [Tabs] Verify/fix layout for safeAreaInsets in Tab Bar View
This was filed as an internal issue. If you are a Googler, please visit [b/136175944](http://b/136175944) for more details.

---
#### Internal data
- Associated internal bug: [b/136175944](http://b/136175944)

### PR DESCRIPTION
**Pull request by leonmz**
Add safe area handling to MDCTabBarView

## Manual testing matrix

|Device|iOS|Justified|Scrollable|
|---|---|---|---|
|iPhone SE|10.3.1|iphonese-10 3 1-justified|iphonese-10 3 1-scrollable|
|iPhone SE|12.0|iphonese-12 0-justified|iphonese-12 0-scrollable|
|iPhone XR|12.0|iphonexr-12 0-justified|iphonexr-12 0-scrollable|
|iPhone X|11.2|iphonex-11 2-justified|iphonex-11 2-scrollable|
|iPad 2|9.3|ipad2-9 3-justified|ipad2-9 3-scrollable|

Closes #7733

### CODE REVIEW

**Comment by leonmz** [comment]
This PR broken the selection visible function for some reason... looking into it now

**Comment by leonmz** [comment]
One bug I found is that iPhone 7 iOS 11.2 the tab bar view cannot scroll.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 1]
```diff
@@ -48,6 +48,13 @@ @interface MDCTabBarView ()
 /** The image tint colors for bar items. */
 @property(nonnull, nonatomic, strong)
     NSMutableDictionary<NSNumber *, UIColor *> *stateToImageTintColor;
+
+/** The constraints that used to set up justified view propertly inseted by safe area. */
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit/grammar: "Constraints unique to the 'Justified' layout style."

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 2]
```diff
@@ -48,6 +48,13 @@ @interface MDCTabBarView ()
...
+@property(nullable, nonatomic) NSArray<NSLayoutConstraint *> *justifiedSafeAreaInsetConstraints;
+
+/** The constraints that used to set up scrollable views */
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit/grammar: "Constraints unique to the 'Scrollable' layout style."

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 3]
```diff
@@ -48,6 +48,13 @@ @interface MDCTabBarView ()
...
+@property(nullable, nonatomic) NSArray<NSLayoutConstraint *> *widthConstraints;
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit/naming: `scrollableLayoutConstraints`.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 4]
```diff
@@ -48,6 +48,13 @@ @interface MDCTabBarView ()
 /** The image tint colors for bar items. */
 @property(nonnull, nonatomic, strong)
     NSMutableDictionary<NSNumber *, UIColor *> *stateToImageTintColor;
+
+/** The constraints that used to set up justified view propertly inseted by safe area. */
+@property(nullable, nonatomic) NSArray<NSLayoutConstraint *> *justifiedSafeAreaInsetConstraints;
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit/naming: `justifiedLayoutConstraints`.

File: components/Tabs/src/TabBarView/MDCTabBarView.m [hunk 5]
```diff
@@ -352,11 +359,29 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
...
+    if (@available(iOS 11.0, *)) {
+      [NSLayoutConstraint deactivateConstraints:self.widthConstraints];
+      [NSLayoutConstraint activateConstraints:self.justifiedSafeAreaInsetConstraints];
+    }
+  } else {
+    self.containerView.distribution = UIStackViewDistributionFillProportionally;
+    // Inset the safe area insets if it is a scrollable view.
+    if (@available(iOS 11.0, *)) {
+      [NSLayoutConstraint deactivateConstraints:self.justifiedSafeAreaInsetConstraints];
+      self.contentInset = UIEdgeInsetsMake(self.contentInset.top, self.safeAreaInsets.left,
```
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: I believe `adjustedContentInset` already includes the `safeAreaInsets` if `contentInsetAdjustmentBehavior` is `.automatic`. Did you not find that to be the case?

Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Some further reading:
https://medium.com/@wailord/the-particulars-of-the-safe-area-and-contentinsetadjustmentbehavior-in-ios-11-9b842018eeaa

**Review by romoore — CHANGES_REQUESTED** [review]
Given the concerns you shared about iPhone 7/iOS 11.2 on Xcode 10.1, I want to make sure we diagnose that bug before merging

**Comment by romoore** [comment]
Going to play with this and see if I can figure out what's going wrong on iPhone 7.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 1 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 3 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 4 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m (hunk 5 above)
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Able to use `safeAreaLayoutGuide` and `contentLayoutGuide` for the appropriate layout styles. Setting the `contentInsetsAdjustmentBehavior` to `.always` seems to be the correct way to get a UIScrollView to position its content inside the safe area.

**Review by romoore — APPROVED** [review]
I think this works well and the result looks pretty good. I'm not seeing what I'd expect when I apply safe area insets to the bottom of the bar, but those are going to be pretty rare in practice.  It may be simpler in the future to migrate to manual layout for the item views.  At this point, however, I'd say the combination of manual testing and snapshot tests are sufficient to cover the common use cases.

**Comment by leonmz** [comment]
Thanks a lot for working on this Rob! Appreciate for your time on investigating on this :)

**Pull request merged**